File: activity/20200805/junhaolin/283.Move_Zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int size = nums.size();
        for (auto it = nums.begin(); it != nums.end();) {
            if (*it == 0)
                it = nums.erase(it);
            else
                it++;
        }
        int delta = size - nums.size();
        while (delta--)
            nums.push_back(0);
    }
};

File: activity/20200805/ninezero/283_move_zeroes.go
func moveZeroes(nums []int)  {
	start := 0
	for _, value := range nums {
		if value != 0 {
			nums[start] = value
			start++
		}
	}

	for i := start; i < len(nums); i++ {
		nums[i] = 0
	}
}

// 另一种解法
func moveZeroes(nums []int)  {
	j := 0
	for i := 0; i < len(nums); i++ {
		// != 0 的时候做交换
		if nums[i] != 0 {
			nums[j], nums[i] = nums[i], nums[j]
			j++
		}
	}
}